discojs/src/dataset/dataset.ts

```diff
@@ -237,6 +238,60 @@ export class Dataset<T> implements AsyncIterable<T> {
   cached(): Dataset<T> {
     return new CachingDataset(this.#content);
   }
+
+  /** Shuffles the Dataset instance within certain window size */
+  shuffle(windowSize: number){
+    if (windowSize < 1){
+      throw new Error("Shuffle window size should be positive");
+    }
+
+    return new Dataset(
+      async function*(this: Dataset<T>){
+        const iter = this[Symbol.asyncIterator]();
+        const buffer: T[] = [];
+
+        while (buffer.length < windowSize){
+          const n = await iter.next();
+          if (n.done) break;
+          buffer.push(n.value);
+        }
+
+        while (buffer.length > 0){
+          const pick = Math.floor(Math.random() * windowSize);
+          const chosen = buffer[pick];
+
+          const n = await iter.next();
+
+          if (n.done){
+            // swap the picked slot with the last slot in the window
+            // remove the last slot in the window
+            const last = buffer.length - 1;
+            [buffer[pick], buffer[last]] = [buffer[last], buffer[pick]];
+            buffer.pop();
```

Comment:
as we `pop` the last element anyway, we can simplify it as (`pop` returns the last element or undef if empty)
```suggestion
            buffer[pick] = buffer.pop() as T; // cast as buffer can't be empty
```